examples-gallery/plot_crane_moving_a_load.py

```diff
@@ -0,0 +1,294 @@
+# start location and and final location of the load.
+anfang = 0.0
+ende = 15.0
+
+# %%
+# Set the initial and final states to form the instance constraints.
+initial_state_constraints = {xc: anfang, xl: anfang, yl: -par_map[l], q: 0.0,
+    uxc: 0.0, uxl: 0.0, uyl: 0.0, u: 0.0}
+final_state_constraints = {xc: ende, xl: ende, yl: -par_map[l], q: 0.0,
+    uxc: 0.0, uxl: 0.0, uyl: 0.0, u: 0.0}
+instance_constraints = (
+) + tuple(
+    xi.subs({t: t0}) - xi_val for xi, xi_val in
+        initial_state_constraints.items()
+) + tuple(
+    xi.subs({t: tf}) - xi_val for xi, xi_val in
+        final_state_constraints.items()
+)
+# %%
+# Forcing h_opty > 0.0 sometimes avoids negative 'solutions'. Here it also
+# seem to help with the convergence of the optimization: Different bounds,
+# e,g, h_opty = (1.e-4, 1.) gives unreasonable results.
+
+bounds = {
+    F: (-20., 20.),
+    xl: (anfang, ende),
+    xc: (anfang, ende),
+    h_opty: (1.e-5, 0.5),
+    }
+
+# Create the optimization problem.
+prob = Problem(
+    obj,
+    obj_grad,
+    EOM,
+    state_symbols,
+    num_nodes,
+    interval_value,
+    known_parameter_map=par_map,
+    instance_constraints=instance_constraints,
+    bounds=bounds,
+    integration_method=methode,
+    )
+# %%
+# Using a 'reasonable' initial guess speeds up the optimization process a lot.
+# 'Bad' initial guesses may even prevent convergence.
+i1 = [(ende - anfang) / num_nodes * i for i in range(num_nodes)]
+i2 = [0. for _ in range(num_nodes)]
+i3 = i1
+i4 = [-par_map[l] for _ in range(num_nodes)]
+i5 = [0. for _ in range(5*num_nodes)]
+i6 = [0.01]
+initial_guess = np.array(i1 + i2 + i3 + i4 + i5 + i6)
+
+# %%
+# Plot the initial guess.
+prob.plot_trajectories(initial_guess)
+
+
+# this allows to change the maximum number of iterations.
+# Standard is 3000.
+prob.add_option('max_iter', 1000)
+# Find the optimal solution.
+for _ in range(1):
+    solution, info = prob.solve(initial_guess)
+    print('message from optimizer:', info['status_msg'])
+    prob.plot_objective_value()
+    initial_guess = solution
+print('Iterations needed', len(prob.obj_value))
+print(f"Objective value {solution[-1]: .3e}")
+
+# %%
+# plot the accuracy of the solution.
+prob.plot_constraint_violations(solution)
+
+# %%
+# Plot the state trajectories.
+fig, ax = plt.subplots(9, 1, figsize=(7.25, 3.0*9), sharex=True)
+prob.plot_trajectories(solution, ax)
+
+# %%
+# Animate the simulation.
+fps = 30
+
+tf = solution[-1] * (num_nodes - 1)
+state_vals, input_vals, _ = parse_free(solution, len(state_symbols),
+    len(specified_symbols), num_nodes)
+t_arr = np.linspace(t0, tf, num_nodes)
+state_sol = CubicSpline(t_arr, state_vals.T)
+input_sol = CubicSpline(t_arr, input_vals.T)
+
+coordinates = P1.pos_from(O).to_matrix(N)
+coordinates = coordinates.row_join(P2.pos_from(O).to_matrix(N))
+
+pL, pL_vals = zip(*par_map.items())
+coords_lam = sm.lambdify((*state_symbols, F, *pL),
+    coordinates, cse=True)
+
+# Dimensions of the mover and the load.
```

Comment:
Code comment, will no render as text.

Reply:
> Code comment, will no render as text.

Should code comments be avoided in general?

Reply:
For these examples, yes, but it is better to have lots of rendered explanatory text so people can easily learn from the examples.